usr/lib/ep11_stdll/ep11_specific.c

```diff
@@ -489,6 +491,394 @@ static CK_RV cleanse_attribute(TEMPLATE *template,
     return CKR_OK;
 }
 
+static CK_RV h_opaque_2_blob(STDLL_TokData_t * tokdata, CK_OBJECT_HANDLE handle,
+                             CK_BYTE ** blob, size_t * blob_len,
+                             OBJECT ** kobj, OBJ_LOCK_TYPE lock_type);
+static CK_RV obj_opaque_2_blob(STDLL_TokData_t *tokdata, OBJECT *key_obj,
+                               CK_BYTE **blob, size_t *blobsize);
+
+static CK_RV check_ab_supported(CK_KEY_TYPE type) {
+    switch(type) {
+    case CKK_AES:
+    case CKK_DES2:
+    case CKK_DES3:
+    case CKK_GENERIC_SECRET:
+    case CKK_RSA:
+    case CKK_EC:
+    case CKK_DSA:
+    case CKK_DH:
+        return CKR_OK;
+    default:
+        TRACE_ERROR("%s key type not supported for ab: 0x%lx\n", __func__, type);
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+}
+
+static CK_RV check_ab_pair(CK_ATTRIBUTE_PTR pPublicKeyTemplate,
+                           CK_ULONG ulPublicKeyAttributeCount,
+                           CK_ATTRIBUTE_PTR pPrivateKeyTemplate,
+                           CK_ULONG ulPrivateKeyAttributeCount)
+{
+    CK_RV rc;
+    CK_BBOOL abpub = FALSE, abpriv = FALSE;
+
+    rc = get_bool_attribute_by_type(pPublicKeyTemplate, ulPublicKeyAttributeCount,
+                                    CKA_IBM_ATTRBOUND, &abpub);
+    if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE) {
+        TRACE_ERROR("Failed to retrieve CKA_IBM_ATTRBOUND attribute from public template. rc=0x%lx\n", rc);
+        return rc;
+    }
+    rc = get_bool_attribute_by_type(pPrivateKeyTemplate, ulPrivateKeyAttributeCount,
+                                    CKA_IBM_ATTRBOUND, &abpriv);
+    if (rc != CKR_OK && rc != CKR_TEMPLATE_INCOMPLETE) {
+        TRACE_ERROR("Failed to retrieve CKA_IBM_ATTRBOUND attribute from private template. rc=0x%lx\n", rc);
+        return rc;
+    }
+    if (abpub != abpriv) {
+        TRACE_ERROR("Only one of the key pair is attribute bound.\n");
+        return CKR_TEMPLATE_INCONSISTENT;
+    }
+    return CKR_OK;
+}
+
+static CK_RV check_ab_kek_type(TEMPLATE *tmpl, CK_RV errval)
+{
+    CK_KEY_TYPE kektype;
+    CK_RV rc;
+
+    rc = template_attribute_get_ulong(tmpl, CKA_KEY_TYPE, &kektype);
+    if (rc == CKR_OK) {
+        if (kektype != CKK_RSA && !is_keytype_symmetric(kektype)) {
+            TRACE_ERROR("KEK key type %ld not supported for AB wrap/unwrap\n",
+                        kektype);
+            return errval;
+        }
+    }
+    return rc;
+}
+
+static CK_RV check_ab_attributes(CK_ATTRIBUTE *attrs,
+                                 CK_ULONG attrs_len,
+                                 CK_OBJECT_CLASS kc)
+{
+    CK_RV rc;
+    CK_BBOOL attrbound = FALSE, sensitive = FALSE;
+
+    if (kc != CKO_PUBLIC_KEY) {
```

Comment:
When checking for (kc == CKO_PUBLIC_KEY) and exiting here, we would save an indentation

Reply:
That would add one more line.  I do not see an advantage here.